Readded Commands.

Waterpicker opened the pull request.

Waterpicker committed:
Added Commands.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/ForestryMC/ForestryMC?pullRequest=2571) <br/>All committers have signed the CLA.

Waterpicker committed:
Fixed formatting

Nedelosk commented on src/main/java/forestry/apiculture/commands/BeeModeHelper.java:
```diff
@@ -1,61 +1,62 @@
+import forestry.core.commands.ICommandModeHelper;
+
+public class BeeModeHelper implements ICommandModeHelper {
+//TODO commands
```
What about all this "//TODO commands" comments?

Nedelosk commented on src/main/java/forestry/apiculture/commands/CommandBeeGive.java:
```diff
@@ -1,175 +1,55 @@
+				.then(Commands.argument("bee", EnumArgument.enumArgument(BeeDefinition.class))
+						.then(Commands.argument("type", EnumArgument.enumArgument(EnumBeeType.class))
+								.then(Commands.argument("player", EntityArgument.player())
+										.executes(a -> execute(a.getSource(), a.getArgument("bee", BeeDefinition.class), a.getArgument("type", EnumBeeType.class), EntityArgument.getPlayer(a, "player"))))
```
What about bees that are added by other mods?

Nedelosk commented on src/main/java/forestry/arboriculture/commands/CommandTreeSpawn.java:
```diff
@@ -1,102 +1,35 @@
+public final class CommandTreeSpawn {
+	public static ArgumentBuilder<CommandSource, ?> register(String name, ITreeSpawner treeSpawner) {
+		return Commands.literal(name).requires(PermLevel.ADMIN)
+				.then(Commands.argument("type", EnumArgument.enumArgument(TreeDefinition.class))
```
What about trees from other mods ?

Nedelosk commented on src/main/java/forestry/core/commands/CommandListAlleles.java:
```diff
@@ -1,59 +1,48 @@
+import genetics.api.individual.IGenome;
+import genetics.api.individual.IIndividual;
+
+public class CommandListAlleles {
```
Would it be a good idea to move this to the genetics mod?

Waterpicker committed:
Refractored some forestry command classes to genetics.

And make BeeGive, ButterflyGive, and Forest/Tree spawn command work with other commands.

Nedelosk commented on src/main/java/forestry/arboriculture/commands/CommandTreeSpawn.java:
```diff
@@ -10,26 +10,76 @@
+		public ITree parse(final StringReader reader) throws CommandSyntaxException {
+			ResourceLocation location = ResourceLocation.read(reader);
+
+			return TreeManager.treeRoot.getIndividualTemplates().stream().filter(a -> a.getGenome().getActiveAllele(TreeChromosomes.SPECIES).getRegistryName().equals(location)).findFirst().orElseThrow(() -> new SimpleCommandExceptionType(new LiteralMessage("Invalid Bee Type: " + location)).create());
```
You can use `TreeManager.treeRoot.templateAsIndividual(TreeManager.treeRoot.getTemplate(location.toString()))`

Waterpicker committed:
Updated Tree Spawn based on suggestion.

Nedelosk commented on src/main/java/forestry/arboriculture/commands/CommandTreeSpawn.java:
```diff
@@ -1,102 +1,82 @@
+
+		@Override
+		public <S> CompletableFuture<Suggestions> listSuggestions(final CommandContext<S> context, final SuggestionsBuilder builder) {
+			return ISuggestionProvider.suggest(TreeManager.treeRoot.getIndividualTemplates().stream()
```
Would a cache be a good idea so we don't have iterate every time over the templates here or is cached by the command system itself?

Nedelosk commented on src/main/java/forestry/apiculture/commands/CommandBeeGive.java:
```diff
@@ -1,175 +1,105 @@
+		public IBee parse(final StringReader reader) throws CommandSyntaxException {
+			ResourceLocation location = ResourceLocation.read(reader);
+
+			return BeeManager.beeRoot.getIndividualTemplates().stream().filter(a -> a.getGenome().getActiveAllele(BeeChromosomes.SPECIES).getRegistryName().equals(location)).findFirst().orElseThrow(() -> new SimpleCommandExceptionType(new LiteralMessage("Invalid Bee Type: " + location)).create());
```
Same as with the trees just with the 'BeeManager.beeRoot'

Waterpicker committed:
Remove Command TODO and old code.
